provisioning/miniprov/hare_mp/cdf.py

```diff
@@ -64,11 +65,15 @@ def _get_cluster_id(self) -> str:
     def _create_node_descriptions(self) -> List[NodeDesc]:
         nodes: List[NodeDesc] = []
         conf = self.provider
+        services = conf.search_val('node', 'services',
```

Comment:
What does this method return? Will it return just keys that match the given arguments?

Reply:
@knekrasov 
Output will be:
['node>aaa120a9e051d103c164f605615c32a4>components[1]>services[0]', 'node>bbb340f79047df9bb52fa460615c32a5>components[1]>services[0]', 'node>ccc8700fe6797ed532e311b0615c32a7>components[1]>services[0]']

Reply:
So it they are the keys, right.

How this code is supposed to work then?
```python
result = list(filter(lambda svc: machine_id in svc, services))
```

Perhaps your intention was to check if `machine_id` is mentioned _within_ the key like `node>aaa120a9e051d103c164f605615c32a4>components[1]>services[0]`, but `in` expression works as a strict equality. Machine ID like `aaa120a9e051d103c164f605615c32a4` will never be equal to `node>aaa120a9e051d103c164f605615c32a4>components[1]>services[0]`

Reply:
Basic use case for filter function : filter(function, iterable)
So here lambda expression "lambda svc: machine_id in svc" as similar as "True if machine_id in svc else False" 
eg. 
True if 'aaa120a9e051d103c164f605615c32a4' in 'node>aaa120a9e051d103c164f605615c32a4>components[1]>services[0]' else False

**Below output will help** :
[root@ssc-vm-g3-rhev4-0864 ~]# python3 test.py
machine_id : aaa120a9e051d103c164f605615c32a4       result: ['node>aaa120a9e051d103c164f605615c32a4>components[1]>services[0]']
machine_id : bbb340f79047df9bb52fa460615c32a5       result: ['node>bbb340f79047df9bb52fa460615c32a5>components[1]>services[0]']
machine_id : ccc8700fe6797ed532e311b0615c32a7       result: ['node>ccc8700fe6797ed532e311b0615c32a7>components[1]>services[0]']
machine_id : eee8700fe6797ed532e311b0615c32a7       result: []
machine_id : fff8700fe6797ed532e311b0615c32a7       result: []
['aaa120a9e051d103c164f605615c32a4', 'bbb340f79047df9bb52fa460615c32a5', 'ccc8700fe6797ed532e311b0615c32a7']

Reply:
Thank you, I understand how `filter` works :-)

My only concern is this lambda: `lambda svc: machine_id in svc, services`

What is the result type of this lambda? When, according to your expectations, this lambda should evaluate to True?